db_test.go

```diff
@@ -0,0 +1,350 @@
+			},
+		},
+		{
+			&Account{},
+			fury.Where,
+			fury.IsGreaterThan("userid", 0),
+			db.Find,
+			&Account{
+				UserID:    1,
+				Username:  "test1",
+				Password:  "test1",
+				Email:     "test1@test.com",
+				CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+			},
+		},
+		{
+			&[]*Account{},
+			fury.Where,
+			fury.IsGreaterThan("account.userid", 0),
+			db.Find,
+			&[]*Account{
+				&Account{
+					UserID:    1,
+					Username:  "test1",
+					Password:  "test1",
+					Email:     "test1@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+				&Account{
+					UserID:    2,
+					Username:  "test2",
+					Password:  "test2",
+					Email:     "test2@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		if err := tc.query(tc.have, tc.options(tc.optionsArgs)); err != nil {
+			t.Error(err)
+		}
+
+		if !reflect.DeepEqual(tc.have, tc.want) {
+			t.Errorf("Error: expected %v, found %v", tc.want, tc.have)
+		}
+	}
+}
+
+func TestInsertQuery(t *testing.T) {
+	type queryFunc func(out interface{}, opts ...fury.QueryOption) error
+	type optionFunc func(conditions interface{}) fury.QueryOption
+
+	cases := []struct {
+		have interface{}
+		want interface{}
+	}{
+		{
+			&Account{
+				UserID: 3,
+			},
+			&Account{
+				UserID:    3,
+				Username:  "test3",
+				Password:  "test3",
+				Email:     "test3@test.com",
+				CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+			},
+		},
+		{
+			&[]*Account{
+				&Account{
+					UserID: 4,
+				},
+				&Account{
+					UserID: 5,
+				},
+			},
+			&[]*Account{
+				&Account{
+					UserID:    4,
+					Username:  "test4",
+					Password:  "test4",
+					Email:     "test4@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+				&Account{
+					UserID:    5,
+					Username:  "test5",
+					Password:  "test5",
+					Email:     "test5@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		if err := db.Insert(tc.want); err != nil {
+			t.Error(err)
+		}
+
+		if err := db.Find(tc.have); err != nil {
+			t.Error(err)
+		}
+
+		if !reflect.DeepEqual(tc.want, tc.have) {
+			t.Errorf("Error: expected %v, found %v", tc.want, tc.have)
+		}
+	}
+}
+
+func TestUpdateQuery(t *testing.T) {
```

Comment:
jika seperti ini jika function ini dijalankan secara parallel dengan function `TestInsertQuery`  atau yang dijalankan hanya `TestUpdateQuery` maka akan error yah?

karena secara default command `go test` akan menjalankan testing secara parallel.
```
by default, it is set to the value of GOMAXPROCS.      Note that -parallel only applies within a single test binary.
     The 'go test' command may run tests for different packages  in parallel as well, according to the setting of the -p flag  (see 'go help build').
```

rule of thumb: setiap test harus ter-isolate & tidak tergantung dengan function lainnya

Reply:
jika ingin setiap test berjalan secara sequence. bisa ubah test nya menjadi seperti ini

```
func TestMainQuery(t *testing.T){
t.Run("insert", testInsertQuery)
t.Run("update", testUpdateQuery)
}
```
jangan lupa setiap functionnnya diubah menjadi private

Reply:
Okeoke, nanti yang ini sekalian ditambahin

Reply:
> rule of thumb: setiap test harus ter-isolate & tidak tergantung dengan function lainnya

Ini yang dimaksud context-nya kan? Maksudnya seperti jangan ada shared state variable. Kalau misalkan untuk test insert, update, delete mesti cek juga dengan find gpp kan ya?

Btw sudah ditambahin seeder-nya dan tiap method gunakan data yang beda kecuali yang find dan first karena hanya select. Kalau sudah begitu, ngga perlu dijadikan sequence kan?

Reply:
iyah bisa ditambahkan feeder untuk setiap test.

Reply:
Oke kalau begitu